src/Symfony/Component/Routing/Generator/UrlGenerator.php

```
@@ -142,6 +142,17 @@ public function generate(string $name, array $parameters = [], int $referenceTyp
      */
     protected function doGenerate(array $variables, array $defaults, array $requirements, array $tokens, array $parameters, string $name, int $referenceType, array $hostTokens, array $requiredSchemes = []): string
     {
+        if (isset($parameters['_query'])) {
+            if (!is_array($parameters['_query'])) {
+                throw new InvalidParameterException('The "_query" parameter must be an array.');
```

what about not throwing and instead fallback to the current behavior?

---

I feel like it is a bit confusing to have different behaviours for `_query`:

```php
$router->generate('...', ['_query' => 'foo']); // ?_query=foo
$router->generate('...', ['_query' => ['foo' => 'bar'])); //?foo=bar
```

And in this sense, I'd prefer to actually *reserve* `_query` for this purpose, and fail hard if you attempt to use it in another way.

On the other hand, doing as you suggest further diminishes the risk of breaking existing applications.

A middle ground could be the following:

- when `_query` is an array: interpret it as a map of query parameters
- when `_query` is a scalar:
  - in Symfony 7.4: interpret it as any other parameter, and trigger a deprecation notice
  - in Symfony 8.0: throw an exception

WDYT?

---

> A middle ground could be the following:

let's do this :+1:

---

Done!